hub/core/tensor.py

```diff
@@ -117,13 +123,51 @@ def extend_tensor(array: np.ndarray, key: str, storage: StorageProvider, **kwarg
         raise TensorDoesNotExistError(key)
 
     tensor_meta, index_meta = _get_metas_from_kwargs(key, storage, **kwargs)
-
     # extend is guaranteed to have a batch axis
     tensor_meta.check_batch_is_compatible(array)
 
     write_array(array, key, storage, tensor_meta, index_meta)
 
 
+def add_index_map_to_tensor(
```

Comment:
this function shouldn't exist. please modify `append_tensor` to accept a `dict` as an alternate input type and try to reduce code copies

you should also remove the `isinstance` check in the API, please migrate this check into `append_tensor`.

Reply:
another note: please remove all `index_map` references and replace them with `index_meta`